File: tools/scripts/split-relations.py
#!/usr/bin/env python3
"""
Split relations.yaml into individual arc-kinds/{family}/{name}.yaml files.

Usage:
    python tools/scripts/split-relations.py

This script:
1. Reads packages/core/models/relations.yaml
2. Creates one YAML file per relation in arc-kinds/{family}/{kebab-name}.yaml
3. Skips relations that already have arc-kind files
"""

import os
import re
import yaml
from pathlib import Path

# Find monorepo root
def find_root() -> Path:
    current = Path(__file__).resolve()
    while current != current.parent:
        if (current / "pnpm-workspace.yaml").exists():
            return current
        current = current.parent
    raise RuntimeError("Could not find monorepo root")

ROOT = find_root()
RELATIONS_PATH = ROOT / "packages/core/models/relations.yaml"
ARC_KINDS_DIR = ROOT / "packages/core/models/arc-kinds"


def to_kebab_case(name: str) -> str:
    """Convert SCREAMING_SNAKE to kebab-case."""
    return name.lower().replace("_", "-")


def format_source_target(value) -> str:
    """Format source/target for YAML output."""
    if isinstance(value, list):
        return value  # Keep as list
    return value  # Keep as string


def infer_scope(source, target) -> str:
    """Infer scope based on source/target patterns."""
    # Self-referential = intra_realm
    if source == target:
        return "intra_realm"
    # Cross-realm if spanning global/project/shared
    return "intra_realm"  # Default for most


def build_cypher_pattern(rel_type: str, source, target, properties: list | None) -> str:
    """Build Cypher pattern string."""
    src_str = source if isinstance(source, str) else source[0]
    tgt_str = target if isinstance(target, str) else target[0]

    if properties:
        prop_names = ", ".join(p if isinstance(p, str) else p.get("name", p) for p in properties)
        return f"({src_str})-[:{rel_type} {{{prop_names}}}]->({tgt_str})"
    return f"({src_str})-[:{rel_type}]->({tgt_str})"


def write_arc_kind(rel: dict, output_path: Path):
    """Write a single arc-kind YAML file."""
    rel_type = rel["type"]
    family = rel["family"]
    source = rel.get("source")
    target = rel.get("target")
    cardinality = rel.get("cardinality", "many_to_many")
    llm_context = rel.get("llm_context", "")
    properties = rel.get("properties")
    inverse_of = rel.get("inverse_of")
    is_self_ref = rel.get("is_self_referential", False)

    # Build arc document
    arc = {
        "name": rel_type,
        "family": family,
        "scope": infer_scope(source, target),
        "source": format_source_target(source),
        "target": format_source_target(target),
        "cardinality": cardinality,
    }

    # Add optional fields
    if is_self_ref:
        arc["is_self_referential"] = True

    if properties:
        # Normalize properties to full format
        arc["properties"] = []
        for prop in properties:
            if isinstance(prop, str):
                arc["properties"].append({
                    "name": prop,
                    "type": "string",
                    "required": False,
                })
            else:
                arc["properties"].append(prop)

    if inverse_of:
        arc["inverse_of"] = inverse_of

    arc["llm_context"] = llm_context
    arc["cypher_pattern"] = build_cypher_pattern(rel_type, source, target, properties)

    # Build full document
    doc = {"arc": arc}

    # Create header comment
    short_desc = llm_context.split(".")[0] if llm_context else f"{rel_type} relationship"
    header = f"""\
# packages/core/models/arc-kinds/{family}/{to_kebab_case(rel_type)}.yaml
# ArcKind: {rel_type} — {short_desc}

"""

    # Write file
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with open(output_path, "w") as f:
        f.write(header)
        yaml.dump(doc, f, default_flow_style=False, allow_unicode=True, sort_keys=False)

    return output_path


def main():
    # Load relations.yaml
    with open(RELATIONS_PATH) as f:
        data = yaml.safe_load(f)

    relations = data.get("relations", [])
    print(f"Found {len(relations)} relations in relations.yaml")

    created = 0
    skipped = 0
    errors = []

    for rel in relations:
        if "type" not in rel:
            continue

        rel_type = rel["type"]
        family = rel.get("family")

        if not family:
            errors.append(f"{rel_type}: missing family")
            continue

        # Determine output path
        filename = f"{to_kebab_case(rel_type)}.yaml"
        output_path = ARC_KINDS_DIR / family / filename

        # Skip if already exists
        if output_path.exists():
            print(f"  SKIP {family}/{filename} (already exists)")
            skipped += 1
            continue

        try:
            write_arc_kind(rel, output_path)
            print(f"  CREATE {family}/{filename}")
            created += 1
        except Exception as e:
            errors.append(f"{rel_type}: {e}")

    print(f"\nSummary:")
    print(f"  Created: {created}")
    print(f"  Skipped: {skipped}")
    print(f"  Errors: {len(errors)}")

    if errors:
        print("\nErrors:")
        for e in errors:
            print(f"  - {e}")


if __name__ == "__main__":
    main()

File: tools/scripts/split-relations.mjs
#!/usr/bin/env node
/**
 * Split relations.yaml into individual arc-kinds/{family}/{name}.yaml files.
 *
 * Usage:
 *   node tools/scripts/split-relations.mjs
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
// Use js-yaml or yaml from core package
// We'll manually install yaml in the script context
// For now, use a simple YAML parser approach
import * as yaml from '../../packages/core/node_modules/yaml/dist/index.js';
const { parse, stringify } = yaml;

const __dirname = dirname(fileURLToPath(import.meta.url));

// Find monorepo root
function findRoot() {
  let current = __dirname;
  while (current !== '/') {
    if (existsSync(join(current, 'pnpm-workspace.yaml'))) {
      return current;
    }
    current = dirname(current);
  }
  throw new Error('Could not find monorepo root');
}

const ROOT = findRoot();
const RELATIONS_PATH = join(ROOT, 'packages/core/models/relations.yaml');
const ARC_KINDS_DIR = join(ROOT, 'packages/core/models/arc-kinds');

function toKebabCase(name) {
  return name.toLowerCase().replace(/_/g, '-');
}

function inferScope(source, target) {
  if (source === target) return 'intra_realm';
  return 'intra_realm';
}

function buildCypherPattern(relType, source, target, properties) {
  const srcStr = Array.isArray(source) ? source[0] : source;
  const tgtStr = Array.isArray(target) ? target[0] : target;

  if (properties && properties.length > 0) {
    const propNames = properties
      .map((p) => (typeof p === 'string' ? p : p.name || p))
      .join(', ');
    return `(${srcStr})-[:${relType} {${propNames}}]->(${tgtStr})`;
  }
  return `(${srcStr})-[:${relType}]->(${tgtStr})`;
}

function writeArcKind(rel, outputPath) {
  const {
    type: relType,
    family,
    source,
    target,
    cardinality = 'many_to_many',
    llm_context: llmContext = '',
    properties,
    inverse_of: inverseOf,
    is_self_referential: isSelfRef,
  } = rel;

  // Build arc document
  const arc = {
    name: relType,
    family,
    scope: inferScope(source, target),
    source,
    target,
    cardinality,
  };

  if (isSelfRef) {
    arc.is_self_referential = true;
  }

  if (properties) {
    arc.properties = properties.map((prop) => {
      if (typeof prop === 'string') {
        return { name: prop, type: 'string', required: false };
      }
      return prop;
    });
  }

  if (inverseOf) {
    arc.inverse_of = inverseOf;
  }

  arc.llm_context = llmContext;
  arc.cypher_pattern = buildCypherPattern(relType, source, target, properties);

  const doc = { arc };

  // Build header comment
  const shortDesc = llmContext
    ? llmContext.split('.')[0]
    : `${relType} relationship`;
  const header = `# packages/core/models/arc-kinds/${family}/${toKebabCase(relType)}.yaml
# ArcKind: ${relType} — ${shortDesc}

`;

  // Ensure directory exists
  mkdirSync(dirname(outputPath), { recursive: true });

  // Write file
  const yamlContent = stringify(doc, { lineWidth: 100 });
  writeFileSync(outputPath, header + yamlContent, 'utf8');

  return outputPath;
}

function main() {
  // Load relations.yaml
  const content = readFileSync(RELATIONS_PATH, 'utf8');
  const data = parse(content);
  const relations = data.relations || [];

  console.log(`Found ${relations.length} relations in relations.yaml`);

  let created = 0;
  let skipped = 0;
  const errors = [];

  for (const rel of relations) {
    if (!rel.type) continue;

    const { type: relType, family } = rel;

    if (!family) {
      errors.push(`${relType}: missing family`);
      continue;
    }

    const filename = `${toKebabCase(relType)}.yaml`;
    const outputPath = join(ARC_KINDS_DIR, family, filename);

    if (existsSync(outputPath)) {
      console.log(`  SKIP ${family}/${filename} (already exists)`);
      skipped++;
      continue;
    }

    try {
      writeArcKind(rel, outputPath);
      console.log(`  CREATE ${family}/${filename}`);
      created++;
    } catch (e) {
      errors.push(`${relType}: ${e.message}`);
    }
  }

  console.log(`\nSummary:`);
  console.log(`  Created: ${created}`);
  console.log(`  Skipped: ${skipped}`);
  console.log(`  Errors: ${errors.length}`);

  if (errors.length > 0) {
    console.log('\nErrors:');
    for (const e of errors) {
      console.log(`  - ${e}`);
    }
  }
}

main();
